remove use of <iterator>

`<iterator>` includes `<iostream>` which leads me to believe we need to ban this header (we don't want to use huge standard library features like iostream in constrained code).

Furthermore, it's breaking the K32W build by bringing in locale support (arguably the problem is in K32W SDK, but we absolutely do not want to bring in localization support).

```
In file included from /home/spang/connectedhomeip2/.environment/cipd/pigweed/arm-none-eabi/include/c++/9.3.1/bits/ios_base.h:41,
                 from /home/spang/connectedhomeip2/.environment/cipd/pigweed/arm-none-eabi/include/c++/9.3.1/ios:42,
                 from /home/spang/connectedhomeip2/.environment/cipd/pigweed/arm-none-eabi/include/c++/9.3.1/ostream:38,
                 from /home/spang/connectedhomeip2/.environment/cipd/pigweed/arm-none-eabi/include/c++/9.3.1/iterator:64,
                 from ../../src/transport/AdminPairingTable.h:25,
                 from ../../src/app/server/Server.h:22,
                 from ../../examples/lock-app/k32w/main/AppTask.cpp:21:
/home/spang/connectedhomeip2/.environment/cipd/pigweed/arm-none-eabi/include/c++/9.3.1/bits/locale_classes.h:299:31: error: constructor cannot be static member function
  299 |     global(const locale& __loc);
      |                               ^
/home/spang/connectedhomeip2/.environment/cipd/pigweed/arm-none-eabi/include/c++/9.3.1/bits/locale_classes.h:298:12: error: 'std::locale::locale(const std::locale&)' cannot be overloaded with 'std::locale::locale(const std::locale&)'
  298 |     static locale
      |            ^~~~~~
/home/spang/connectedhomeip2/.environment/cipd/pigweed/arm-none-eabi/include/c++/9.3.1/bits/locale_classes.h:126:5: note: previous declaration 'std::locale::locale(const std::locale&)'
  126 |     locale(const locale& __other) throw();
      |     ^~~~~~
```

(the compile failure is caused by `#define global` in K32W SDK headers - @doru91)

--- a/src/lib/mdns/minimal/responders/QueryResponder.h
+++ b/src/lib/mdns/minimal/responders/QueryResponder.h
@@ -164,8 +164,8 @@ class QueryResponderIterator
 {
 public:
     using value_type = QueryResponderRecord;
-    using pointer = QueryResponderRecord*;
-    using reference = QueryResponderRecord&;
+    using pointer    = QueryResponderRecord *;
+    using reference  = QueryResponderRecord &;
 
     QueryResponderIterator() : mCurrent(nullptr), mRemaining(0) {}
     QueryResponderIterator(QueryResponderRecordFilter * recordFilter, Internal::QueryResponderInfo * pos, size_t size) :

--- a/src/lib/mdns/minimal/responders/QueryResponder.h
+++ b/src/lib/mdns/minimal/responders/QueryResponder.h
@@ -22,8 +22,6 @@
 
 #include <inet/InetLayer.h>
 
-#include <iterator>
-
 namespace mdns {
 namespace Minimal {
 
@@ -162,9 +160,13 @@ private:
 
 /// Iterates over an array of QueryResponderRecord items, providing only 'valid' ones, where
 /// valid is based on the provided filter.
-class QueryResponderIterator : public std::iterator<std::input_iterator_tag, QueryResponderRecord>
+class QueryResponderIterator
 {
 public:
+    using value_type = QueryResponderRecord;
+    using pointer = QueryResponderRecord*;
+    using reference = QueryResponderRecord&;
+
     QueryResponderIterator() : mCurrent(nullptr), mRemaining(0) {}
     QueryResponderIterator(QueryResponderRecordFilter * recordFilter, Internal::QueryResponderInfo * pos, size_t size) :
         mFilter(recordFilter), mCurrent(pos), mRemaining(size)

--- a/src/transport/AdminPairingTable.h
+++ b/src/transport/AdminPairingTable.h
@@ -22,7 +22,6 @@
 #pragma once
 
 #include <core/CHIPPersistentStorageDelegate.h>
-#include <iterator>
 #include <support/DLLUtil.h>
 #include <transport/raw/MessageHeader.h>
 
@@ -115,9 +114,13 @@ private:
 /**
  * Iterates over valid admins within a list
  */
-class ConstAdminIterator : public std::iterator<std::forward_iterator_tag, const AdminPairingInfo>
+class ConstAdminIterator
 {
 public:
+    using value_type = AdminPairingInfo;
+    using pointer = AdminPairingInfo*;
+    using reference = AdminPairingInfo&;
+
     ConstAdminIterator(const AdminPairingInfo * start, size_t index, size_t maxSize) :
         mStart(start), mIndex(index), mMaxSize(maxSize)
     {

--- a/src/transport/AdminPairingTable.h
+++ b/src/transport/AdminPairingTable.h
@@ -118,8 +118,8 @@ class ConstAdminIterator
 {
 public:
     using value_type = AdminPairingInfo;
-    using pointer = AdminPairingInfo*;
-    using reference = AdminPairingInfo&;
+    using pointer    = AdminPairingInfo *;
+    using reference  = AdminPairingInfo &;
 
     ConstAdminIterator(const AdminPairingInfo * start, size_t index, size_t maxSize) :
         mStart(start), mIndex(index), mMaxSize(maxSize)

